Cabal/src/Distribution/Simple/BuildTarget.hs

```diff
@@ -246,19 +246,12 @@ reportUserBuildTargetProblems verbosity problems = do
   case [target | UserBuildTargetUnrecognised target <- problems] of
     [] -> return ()
     target ->
-      die' verbosity $
-        unlines
-          [ "Unrecognised build target '" ++ name ++ "'."
-          | name <- target
-          ]
-          ++ "Examples:\n"
-          ++ " - build foo          -- component name "
-          ++ "(library, executable, test-suite or benchmark)\n"
-          ++ " - build Data.Foo     -- module name\n"
-          ++ " - build Data/Foo.hsc -- file name\n"
-          ++ " - build lib:foo exe:foo   -- component qualified by kind\n"
-          ++ " - build foo:Data.Foo      -- module qualified by component\n"
-          ++ " - build foo:Data/Foo.hsc  -- file qualified by component"
+      dieWithException verbosity $
+        UnrecognisedBuildTarget $
+          unlines
```

Comment:
I'm wondering if this processing belongs in the error handler rather than here. It looks like the previous error handling commit did things that way?

Reply:
Yes, in the previous commit this processing has been done in the error handling. I have changed this instance accordingly. 
However for the constructors "ReportBuildTargetProblems", "UnknownBuildTarget","AmbiguousBuildTarget" I retained the processing in the module itself because there was a cyclic module dependency error when I imported them in the error handler.

Reply:
"cyclic module dependency error when I imported them in the error handler."

when you imported precisely what types from what modules in what other module?

Reply:
I tried to import the type `UserBuildTarget` from `BuildTarget.hs` into `Errors.hs`

I was planning to do this "ReportBuildTargetProblems [(UserBuildTarget, [String], String)]" and then do all the processing for the error string in the `Errors.hs` module.

Reply:
I propose that for those constructors, UserBuildTarget is not added, but the `String` result of calling `showUserBuildTarget` on it is, which should suffice. That way the rest of the string processing of the error message can be done in Errors.hs, in keeping with elsewhere.

Reply:
map (\(x,y,z) -> (showUserBuildTarget x, y, z)) targets

Reply:
Thanks, 
I have brought in this change for the constructors "ReportBuildTargetProblems","UnknownBuildTarget","AmbiguousBuildTarget"